File: python/tests/test_config.py
"""Tests for ClientConfig."""

from __future__ import annotations

import pytest

from ensoul.config import (
    API_VERSION,
    DEFAULT_BASE_URL,
    DEFAULT_MAX_RETRIES,
    DEFAULT_TIMEOUT,
    ClientConfig,
)


class TestClientConfig:
    def test_defaults(self):
        config = ClientConfig()
        assert config.base_url == DEFAULT_BASE_URL
        assert config.api_key is None
        assert config.bearer_token is None
        assert config.timeout == DEFAULT_TIMEOUT
        assert config.max_retries == DEFAULT_MAX_RETRIES
        assert config.custom_headers == {}

    def test_custom_values(self):
        config = ClientConfig(
            base_url="https://custom.example.com",
            api_key="sk_test_abc",
            bearer_token="bearer_xyz",
            timeout=60.0,
            max_retries=5,
            custom_headers={"X-Custom": "value"},
        )
        assert config.base_url == "https://custom.example.com"
        assert config.api_key == "sk_test_abc"
        assert config.bearer_token == "bearer_xyz"
        assert config.timeout == 60.0
        assert config.max_retries == 5
        assert config.custom_headers == {"X-Custom": "value"}

    def test_api_url_property(self):
        config = ClientConfig(base_url="https://api.ensoul.ai")
        assert config.api_url == f"https://api.ensoul.ai/{API_VERSION}"

    def test_api_url_strips_trailing_slash(self):
        config = ClientConfig(base_url="https://api.ensoul.ai/")
        assert config.api_url == f"https://api.ensoul.ai/{API_VERSION}"
        # should not have double slashes
        assert "//" not in config.api_url.replace("https://", "")

    def test_api_url_with_custom_base(self):
        config = ClientConfig(base_url="http://localhost:8000")
        assert config.api_url == f"http://localhost:8000/{API_VERSION}"

File: cpp/tests/test_config.cpp
#include <catch2/catch_test_macros.hpp>
#include "ensoul/config.hpp"

using namespace ensoul;

TEST_CASE("ClientConfig has correct defaults", "[config]") {
    ClientConfig config;
    CHECK(config.base_url == "https://api.ensoul-ai.com");
    CHECK(config.timeout_ms == 30000);
    CHECK(config.max_retries == 2);
    CHECK(config.api_key.empty());
    CHECK(config.bearer_token.empty());
    CHECK(config.custom_headers.empty());
}

TEST_CASE("ClientConfig custom values", "[config]") {
    ClientConfig config;
    config.base_url = "https://custom.api.com";
    config.api_key = "my-key";
    config.timeout_ms = 5000;
    config.max_retries = 5;
    CHECK(config.base_url == "https://custom.api.com");
    CHECK(config.api_key == "my-key");
    CHECK(config.timeout_ms == 5000);
    CHECK(config.max_retries == 5);
}

TEST_CASE("api_url strips trailing slash and appends /v1", "[config]") {
    ClientConfig config;
    config.base_url = "https://api.ensoul-ai.com/";
    CHECK(config.api_url() == "https://api.ensoul-ai.com/v1");
}

TEST_CASE("api_url without trailing slash", "[config]") {
    ClientConfig config;
    config.base_url = "https://api.ensoul-ai.com";
    CHECK(config.api_url() == "https://api.ensoul-ai.com/v1");
}

TEST_CASE("Constants have expected values", "[config]") {
    CHECK(std::string(DEFAULT_BASE_URL) == "https://api.ensoul-ai.com");
    CHECK(DEFAULT_TIMEOUT_MS == 30000);
    CHECK(DEFAULT_MAX_RETRIES == 2);
    CHECK(std::string(API_VERSION) == "v1");
}
